apps/shared/avifjpeg.c

```diff
@@ -354,24 +383,134 @@ avifBool avifJPEGRead(const char * inputFilename,
+                foundStandardXMP = AVIF_TRUE;
+            }
+        }
+
+        const avifROData tagExtendedXMP = { (const uint8_t *)"http://ns.adobe.com/xmp/extension/\0", 35 };
+        avifBool foundExtendedXMP = AVIF_FALSE;
+        const size_t extendedXMPGUIDLength = 32;
+        uint8_t extendedXMPGUID[32] = { 0 }; // The value is common to all extended XMP segments.
+        for (jpeg_saved_marker_ptr marker = cinfo.marker_list; marker != NULL; marker = marker->next) {
+            if ((marker->marker == (JPEG_APP0 + 1)) && (marker->data_length > tagExtendedXMP.size) &&
+                !memcmp(marker->data, tagExtendedXMP.data, tagExtendedXMP.size)) {
+                if (!foundStandardXMP) {
+                    fprintf(stderr, "XMP extraction failed: extended XMP segment found, missing standard XMP segment\n");
+                    goto cleanup;
+                }
+
+                // Offset in marker->data segment (skip tag + guid + size + offset).
+                const size_t offsetTillExtendedXMP = tagExtendedXMP.size + extendedXMPGUIDLength + 4 + 4;
+                if (marker->data_length < offsetTillExtendedXMP) {
+                    fprintf(stderr, "XMP extraction failed: truncated extended XMP segment\n");
+                    goto cleanup;
+                }
+                const uint8_t * guid = &marker->data[tagExtendedXMP.size];
+                for (size_t c = 0; c < extendedXMPGUIDLength; ++c) {
+                    // According to Adobe XMP Specification Part 3 section 1.1.3.1:
+                    //   "128-bit GUID stored as a 32-byte ASCII hex string, capital A-F, no null termination"
+                    if (((guid[c] < '0') || (guid[c] > '9')) && ((guid[c] < 'A') || (guid[c] > 'F'))) {
+                        fprintf(stderr, "XMP extraction failed: invalid XMP segment GUID\n");
+                        goto cleanup;
+                    }
+                }
+                // Size of the current extended segment.
+                const size_t extendedXMPSize = marker->data_length - offsetTillExtendedXMP;
+                // Expected size of the sum of all extended segments.
+                // According to Adobe XMP Specification Part 3 section 1.1.3.1:
+                //   "full length of the ExtendedXMP serialization as a 32-bit unsigned integer"
+                const uint32_t totalExtendedXMPSize =
+                    avifJPEGReadUint32BigEndian(&marker->data[tagExtendedXMP.size + extendedXMPGUIDLength]);
+                // Offset in totalXMP after standardXMP.
+                // According to Adobe XMP Specification Part 3 section 1.1.3.1:
+                //   "offset of this portion as a 32-bit unsigned integer"
+                const uint32_t extendedXMPOffset =
+                    avifJPEGReadUint32BigEndian(&marker->data[tagExtendedXMP.size + extendedXMPGUIDLength + 4]);
+                if ((uint64_t)standardXMP.size + totalExtendedXMPSize > SIZE_MAX) {
```

Comment:
I usually check for potential overflow before adding two numbers. Casting the numbers to a larger integer type also works. You used that method here and at line 445 below.

When used here, it is more work to verify the correctness. The reason is that in 64-bit builds, `standardXMP.size` (a `size_t`) is already the same size as `uint64_t`, so the `(uint64_t)` cast is a no-op. The correctness comes from the fact that the value stored in `standardXMP.size` is less than `marker->data_length`, which is an `unsigned int`. This takes some work to discover.

```
struct jpeg_marker_struct {
  jpeg_saved_marker_ptr next;   /* next in list, or NULL */
  UINT8 marker;                 /* marker code: JPEG_COM, or JPEG_APP0+n */
  unsigned int original_length; /* # bytes of data in the file */
  unsigned int data_length;     /* # bytes of data saved at data[] */
  JOCTET *data;                 /* the data contained in the marker */
  /* the marker length word is not counted in data_length or original_length */
};
```

It would be good to add an assertion to indicate we are depending on this fact:

```
                    assert(standardXMP.size <= UINT32_MAX);
```

Or replace `standardXMP` with a pair of variables:

```
        const uint8_t *standardXMPData = NULL;
        uint32_t standardXMPSize = 0; // At most 64kB as defined by Adobe XMP Specification Part 3.
        for (jpeg_saved_marker_ptr marker = cinfo.marker_list; marker != NULL; marker = marker->next) {
            if ((marker->marker == (JPEG_APP0 + 1)) && (marker->data_length > tagStandardXMP.size) &&
                !memcmp(marker->data, tagStandardXMP.data, tagStandardXMP.size)) {
                if (foundStandardXMP) {
                    fprintf(stderr, "XMP extraction failed: invalid multiple XMP segments\n");
                    goto cleanup;
                }
                standardXMPData = marker->data + tagStandardXMP.size;
                standardXMPSize = marker->data_length - tagStandardXMP.size;
                foundStandardXMP = AVIF_TRUE;
            }
        }
```